nemo/collections/asr/models/ssl_models.py

```diff
@@ -589,3 +606,407 @@ def multi_validation_epoch_end(self, outputs, dataloader_idx: int = 0):
+            self.encoder.pre_encode = self.pre_encoder
+
+    @property
+    def input_types(self) -> Optional[Dict[str, NeuralType]]:
+        if hasattr(self.preprocessor, '_sample_rate'):
+            input_signal_eltype = AudioSignal(freq=self.preprocessor._sample_rate)
+        else:
+            input_signal_eltype = AudioSignal()
+        return {
+            "input_signal": NeuralType(('B', 'T'), input_signal_eltype, optional=True),
+            "input_signal_length": NeuralType(tuple('B'), LengthsType(), optional=True),
+            "processed_signal": NeuralType(('B', 'D', 'T'), SpectrogramType(), optional=True),
+            "processed_signal_length": NeuralType(tuple('B'), LengthsType(), optional=True),
+            "targets": NeuralType(('B', 'T'), LabelsType(), optional=True),
+            "target_lengths": NeuralType(tuple('B'), LengthsType(), optional=True),
+            "apply_mask": NeuralType(optional=True),
+        }
+
+    @property
+    def output_types(self) -> Optional[Dict[str, NeuralType]]:
+        if self.cfg.num_books == 1 and self.cfg.squeeze_single:
+            logprobs = NeuralType(('B', 'T', 'C'), LogprobsType())
+            tokens = NeuralType(('B', 'T'), LabelsType())
+        else:
+            logprobs = NeuralType(('B', 'T', 'C', 'H'), LogprobsType())
+            tokens = NeuralType(('B', 'T', 'H'), LabelsType())
+        return {
+            "logprobs": logprobs,
+            "encoded_len": NeuralType(tuple('B'), LengthsType()),
+            "masks": NeuralType(('B', 'D', 'T'), SpectrogramType()),
+            "tokens": tokens,
+        }
+
+    @typecheck()
+    def forward(
+        self,
+        input_signal=None,
+        input_signal_length=None,
+        processed_signal=None,
+        processed_signal_length=None,
+        apply_mask=False,
+    ):
+        has_input_signal = input_signal is not None and input_signal_length is not None
+        has_processed_signal = processed_signal is not None and processed_signal_length is not None
+        if (has_input_signal ^ has_processed_signal) == False:
+            raise ValueError(
+                f"{self} Arguments ``input_signal`` and ``input_signal_length`` are mutually exclusive "
+                " with ``processed_signal`` and ``processed_signal_len`` arguments."
+            )
+
+        if not has_processed_signal:
+            processed_signal, processed_signal_length = self.preprocessor(
+                input_signal=input_signal,
+                length=input_signal_length,
+            )
+
+        if self.pre_encoder is not None:
+            # mask after convolutional sub-sampling
+            self.pre_encoder.set_masking_enabled(apply_mask=apply_mask)
+            encoded, encoded_len = self.encoder(audio_signal=processed_signal, length=processed_signal_length)
+            masks = self.pre_encoder.get_current_mask()
+            feats = self.pre_encoder.get_current_feat()
```

Comment:
they are applied at `https://github.com/stevehuang52/NeMo/blob/ssl_pr/nemo/collections/asr/modules/ssl_modules/masking.py#L177`

Reply:
instead of apply mask in the from class variable, can we make it to pass mask as an argument to .get_current_feat(). Better for readability and avoids silent breaking of things